### PR TITLE
feat: Eng-384 convert apiId settings actions to tRPC routes

### PR DESCRIPTION
**Pull request by MichaelUnkey**
## What does this PR do?

Fixes # (issue)

eng-384

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Test api settings to ensure proper functionality. name change and apiWhitelist work as intended. 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

### CODE REVIEW

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx [hunk 1]
```diff
@@ -26,25 +27,39 @@ type Props = {
 export const UpdateApiName: React.FC<Props> = ({ api }) => {
   const { toast } = useToast();
   const { pending } = useFormStatus();
+  const router = useRouter();
+  const updateName = trpc.apiSettings.updateName.useMutation({
+    onSuccess: (_data) => {
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:32]: if we don't use it, we can just remove it
```suggestion
    onSuccess: () => {
```

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx [hunk 2]
```diff
@@ -24,19 +33,36 @@ type Props = {
...
+      name: api.name,
+      apiId: api.id,
+      workspaceId: api.workspaceId,
+    },
+  });
 
+  async function onSubmit(values: z.infer<typeof formSchema>) {
+    try {
+      setLoading(true);
+      await updateName.mutate({
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:50]: actually there's a nicer way to do this

we can move all the UI updating stuff into the `useMutation` function

```ts

  const updateName = trpc.api.create.useMutation({
    onSuccess(){
      toast.success("Your API has been renamed");
      router.refresh();
    },
    onError(err) {
      console.error(err);
      toast.error(err.message);
    },
  });
  ```
  
  and then you can probably remove the `onSubmit` entirely and pass in `updateName.mutate` directly:
  ```ts
  form onSubmit={form.handleSubmit(updateName.mutate)}>
    
  ```

[the remark this answers is not in the document]
Comment (MichaelUnkey) [reply apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx]: Changes made based on your comments. Let me know if it looks ok.

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx [hunk 3]
```diff
@@ -24,19 +32,34 @@ type Props = {
 };
 
 export const UpdateApiName: React.FC<Props> = ({ api }) => {
-  const { pending } = useFormStatus();
+  const { formState } = useForm();
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:35]: we already have a `useForm()` below which we can use
```suggestion
```

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx [hunk 4]
```diff
@@ -24,19 +32,34 @@ type Props = {
...
+  const { isSubmitting } = formState;
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:36]: you can move this down and read it from `form.formState.isSubmitting`

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx [hunk 5]
```diff
@@ -33,22 +41,36 @@ type Props = {
 };
 
 export const UpdateIpWhitelist: React.FC<Props> = ({ api, workspace }) => {
-  const { pending } = useFormStatus();
-
+  const { formState } = useForm();
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx:44]: same as above

File: apps/web/lib/trpc/routers/api.ts [hunk 6]
```diff
@@ -59,21 +59,25 @@ export const apiRouter = t.router({
...
         // For local development when contributors don't have access to the trigger.dev account
-        await db
-          .update(schema.apis)
-          .set({ deletedAt: new Date() })
-          .where(eq(schema.apis.id, input.apiId));
-        await db
-          .update(schema.keys)
-          .set({ deletedAt: new Date() })
-          .where(eq(schema.keys.keyAuthId, api.keyAuthId!));
+        await db.transaction(async (tx) => {
```
Comment (chronark) [inline apps/web/lib/trpc/routers/api.ts:71]: let's do a quick huddle on this later, hit me up

File: apps/web/lib/trpc/routers/api.ts [hunk 7]
```diff
@@ -126,4 +135,109 @@ export const apiRouter = t.router({
...
+          code: "NOT_FOUND",
+        });
+      }
+      const api = ws.apis.find((api) => api.id === input.apiId);
+
+      if (!api) {
+        throw new TRPCError({ message: "api not found", code: "NOT_FOUND" });
+      }
+
+      const _res = await db.transaction(async (tx) => {
```
Comment (chronark) [inline apps/web/lib/trpc/routers/api.ts:170]: we need to add an audit log here

**Review by chronark — CHANGES_REQUESTED** [review]

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx [hunk 8]
```diff
@@ -24,19 +32,31 @@ type Props = {
...
+    },
+    onError(err) {
+      console.log(err);
+      toast.error(err.message);
+    },
+  });
+  async function onSubmit(values: z.infer<typeof formSchema>) {
+    updateName.mutateAsync(values);
+  }
+  const router = useRouter();
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:57]: does this work? we're importing it here, but already use it in line 47

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx [hunk 9]
```diff
@@ -33,22 +41,34 @@ type Props = {
 };
 
 export const UpdateIpWhitelist: React.FC<Props> = ({ api, workspace }) => {
-  const { pending } = useFormStatus();
-
   const isEnabled = workspace.plan === "enterprise";
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx:44]: coordinate with @domeccleston for using the feature flag instead

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx [hunk 10]
```diff
@@ -33,22 +41,34 @@ type Props = {
...
+    onError(err) {
+      console.log(err);
+      toast.error(err.message);
+    },
+  });
+
+  async function onSubmit(values: z.infer<typeof formSchema>) {
+    updateIps.mutateAsync(values);
+  }
+  const router = useRouter();
```
Comment (chronark) [inline apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx:68]: same thing, it looks weird that this is below where we use it

File: apps/web/lib/trpc/routers/api.ts [hunk 11]
```diff
@@ -59,21 +59,46 @@ export const apiRouter = t.router({
...
+            .set({ state: "DELETION_IN_PROGRESS" })
+            .where(eq(schema.apis.id, input.apiId));
+          await tx.insert(schema.auditLogs).values({
+            id: newId("auditLog"),
+            time: new Date(),
+            workspaceId: api.workspaceId,
+            actorType: "user",
+            actorId: ctx.user.id,
+            event: "api.delete",
+            description: `API ${api.name} deleted`,
```
Comment (chronark) [inline apps/web/lib/trpc/routers/api.ts:74]: My bad, this one doesn't delete, but only start the process.

```suggestion
            description: `API ${api.name} marked for deletion`,
```

File: apps/web/lib/trpc/routers/api.ts [hunk 12]
```diff
@@ -59,21 +59,46 @@ export const apiRouter = t.router({
...
+            apiId: api.id,
+          }); // TODO
```
Comment (chronark) [inline apps/web/lib/trpc/routers/api.ts:76]:
```suggestion
          });
```

**Review by chronark — CHANGES_REQUESTED** [review]

File: apps/web/lib/trpc/routers/api.ts [hunk 13]
```diff
@@ -126,4 +156,119 @@ export const apiRouter = t.router({
...
+          code: "NOT_FOUND",
+        });
+      }
+      const api = ws.apis.find((api) => api.id === input.apiId);
+
+      if (!api) {
+        throw new TRPCError({ message: "api not found", code: "NOT_FOUND" });
+      }
+
+      const _res = await db.transaction(async (tx) => {
```
Comment (chronark) [inline apps/web/lib/trpc/routers/api.ts:191]:
```suggestion
      await db.transaction(async (tx) => {
```

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx (hunk 8 above)
Comment (MichaelUnkey) [reply apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-api-name.tsx:57]: Thought I moved this but must have missed it on this file.

File: apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx (hunk 10 above)
Comment (MichaelUnkey) [reply apps/web/app/(authenticated)/(app)/app/apis/[apiId]/settings/update-ip-whitelist.tsx:68]: Moved

File: apps/web/lib/trpc/routers/api.ts (hunk 11 above)
Comment (MichaelUnkey) [reply apps/web/lib/trpc/routers/api.ts:74]: Changed

File: apps/web/lib/trpc/routers/api.ts (hunk 13 above)
Comment (MichaelUnkey) [reply apps/web/lib/trpc/routers/api.ts:191]: Changed

**Review by chronark — APPROVED** [review]

**Pull request merged**